src/test/java/com/coveo/pushapiclient/DocumentUploadQueueTest.java

```diff
@@ -0,0 +1,192 @@
+        String twoMegaByteData = generateStringFromBytes(2 * oneMegaByte);
+
+        documentToAdd = new DocumentBuilder(
+                "https://my.document.uri?ref=1",
+                "My new document")
+                .withData(twoMegaByteData);
+
+        documentToDelete = new DeleteDocument("https://my.document.uri?ref=3");
+
+        closeable = MockitoAnnotations.openMocks(this);
+    }
+
+    @After
+    public void closeService() throws Exception {
+        closeable.close();
+    }
+
+    @Test
+    public void testIsEmpty() throws IOException, InterruptedException {
+        assertTrue(queue.isEmpty());
+    }
+
+    @Test
+    public void testIsNotEmpty() throws IOException, InterruptedException {
+        queue.add(documentToAdd);
+        assertFalse(queue.isEmpty());
+    }
+
+    @Test
+    public void testShouldReturnBatch() throws IOException, InterruptedException {
+        BatchUpdate batchUpdate = new BatchUpdate(
+                new ArrayList<>() {
+                    {
+                        add(documentToAdd);
+                    }
+                }, new ArrayList<>() {
+                    {
+                        add(documentToDelete);
+                    }
+                });
+        queue.add(documentToAdd);
+        queue.add(documentToDelete);
+
+        assertEquals(batchUpdate, queue.getBatch());
+    }
+
+    @Test
+    public void testFlushShouldNotUploadDocumentaWhenRequiredSizeIsNotMet() throws IOException, InterruptedException {
+        queue.add(documentToAdd);
+        queue.add(documentToDelete);
+
+        verify(uploadStrategy, times(0)).apply(any(BatchUpdate.class));
+    }
+
+    @Test
+    public void testShouldAutomaticallyFlushAccumulatedDocuments() throws IOException, InterruptedException {
+        DocumentBuilder firstBulkyDocument = generateDocumentFromSize(2 * oneMegaByte);
+        DocumentBuilder secondBulkyDocument = generateDocumentFromSize(2 * oneMegaByte);
+        DocumentBuilder thirdBulkyDocument = generateDocumentFromSize(2 * oneMegaByte);
+        ArrayList<DeleteDocument> emptyList = new ArrayList<>();
+        BatchUpdate firstBatch = new BatchUpdate(
+                new ArrayList<>() {
+                    {
+                        add(firstBulkyDocument);
+                        add(secondBulkyDocument);
+                    }
+                }, emptyList);
+
+        // Adding 3 documents of 2MB to the queue. After adding the first 2 documents,
+        // the queue size will reach 6MB, which exceeds the maximum queue size
+        // limit. Therefore, the 2 first added documents will automatically be uploaded
+        // to the source.
+        queue.add(firstBulkyDocument);
+        queue.add(secondBulkyDocument);
+
+        // The 3rd document added to the queue will be included in a separate batch,
+        // which will not be uploaded unless the `flush()` method is called or until the
+        // queue size limit has been reached
```

Comment:
is there a way to check that this hasn't been uploaded?

Reply:
Yes, I will add a `verify` to ensure no batch was uplaoded